WalletWasabi/Wallets/TransactionFeeProvider.cs

```diff
@@ -0,0 +1,131 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.Blockchain.TransactionProcessing;
+using WalletWasabi.Extensions;
+using WalletWasabi.Logging;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.Tor.Http.Extensions;
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class TransactionFeeProvider : BackgroundService
+{
+	private readonly TimeSpan _maximumDelay = TimeSpan.FromMinutes(2);
+
+	public TransactionFeeProvider(WasabiHttpClientFactory httpClientFactory)
+	{
+		HttpClient = httpClientFactory.NewHttpClient(httpClientFactory.BackendUriGetter, Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+	}
+
+	public event EventHandler<EventArgs>? RequestedFeeArrived;
+
+	public ConcurrentDictionary<uint256, Money> FeeCache { get; } = new();
+	public ConcurrentQueue<uint256> Queue { get; } = new();
+	private SemaphoreSlim Semaphore { get; } = new(initialCount: 0);
+
+	private IHttpClient HttpClient { get; }
+
+	private async Task FetchTransactionFeeAsync(uint256 txid, CancellationToken cancellationToken)
+	{
+		const int MaxAttempts = 3;
+
+		for (int i = 0; i < MaxAttempts; i++)
+		{
+			try
+			{
+				using CancellationTokenSource timeoutCts = new(TimeSpan.FromSeconds(20));
+				using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
+
+				var response = await HttpClient.SendAsync(
+					HttpMethod.Get,
+					$"api/v{Helpers.Constants.BackendMajorVersion}/btc/Blockchain/get-transaction-fee?transactionId={txid}",
+					null,
+					linkedCts.Token).ConfigureAwait(false);
+
+				if (response.StatusCode != HttpStatusCode.OK)
+				{
+					await response.ThrowRequestExceptionFromContentAsync(cancellationToken).ConfigureAwait(false);
+				}
+
+				Money fee = await response.Content.ReadAsJsonAsync<Money>().ConfigureAwait(false);
+
+				if (!FeeCache.TryAdd(txid, fee))
+				{
+					throw new InvalidOperationException($"Failed to cache {txid} with fee: {fee}");
+				}
+
+				RequestedFeeArrived?.Invoke(this, EventArgs.Empty);
+				return;
+			}
+			catch (Exception ex)
+			{
+				if (cancellationToken.IsCancellationRequested)
+				{
+					return;
+				}
+
+				Logger.LogWarning($"Attempt: {i}. Failed to fetch transaction fee. {ex}");
+			}
+		}
+	}
+
+	public bool TryGetFeeFromCache(uint256 txid, [NotNullWhen(true)] out Money? fee)
+	{
+		return FeeCache.TryGetValue(txid, out fee);
+	}
+
+	public void WalletRelevantTransactionProcessed(object? sender, ProcessedResult e)
+	{
+		if (!e.Transaction.Confirmed && e.Transaction.ForeignInputs.Any())
+		{
+			Queue.Enqueue(e.Transaction.GetHash());
```

Comment:
I fail to realise what benefit we gain if I use `Channels` here instead of a `Queue`. 
Especially that there is an improvement idea to use `PriorityQueue` in case the user opens the transaction details (so that request goes out sooner)

Please elaborate.

Reply:
According to my experience the more concurrent items we have the harder the maintenance and the initial review as well. So that is why I suggested having only one object - the channel - as the interconnection between the loop and the "producer" method. Everything else can happen inside the loop, even prioritization. Both solutions will work it is up to you. I suggest to try it to learn and you can decide which you like better. We can pair program it.